src/axolotl/utils/config/models/input/v0_4_1/__init__.py

```diff
@@ -0,0 +1,931 @@
+            )
+        return root
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_pretraining_w_max_steps(cls, data):
+        if data.get("pretraining_dataset") and not data.get("max_steps"):
+            raise ValueError(
+                "max_steps must be set when using iterable pretraining_dataset, Trainer can't infer length and schedule optimizer/learning rate without it!"
+            )
+        return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_pretraining_w_group_by_length(cls, data):
+        if data.get("pretraining_dataset") and data.get("group_by_length"):
+            LOG.warning(
+                "You probably want to disable group_by_length as it will force a streamed dataset to download completely."
+            )
+        return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_gptq_w_revision(cls, data):
+        if data.get("gptq") and data.get("model_revision"):
+            raise ValueError(
+                "model_revision is not supported for GPTQ models. "
+                + "Please download the model from HuggingFace Hub manually for correct branch, "
+                + "point to its path, and remove model_revision from the config."
+            )
+        return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_sample_packing_w_xformers(cls, root):
+        if root.get("sample_packing") and root.get("xformers_attention"):
+            raise ValueError(
+                "sample_packing not compatible with xformers_attention. Use flash_attention"
+            )
+
+        return root
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_sample_packing_w_sdpa_bf16(cls, root):
+        if (
+            root.get("sample_packing")
+            and root.get("sdp_attention")
+            and (root.get("bfloat16") or root.get("bf16"))
+        ):
+            # https://github.com/pytorch/pytorch/blob/1b03423526536b5f3d35bdfa95ccc6197556cf9b/test/test_transformers.py#L2440-L2450
+            LOG.warning(
+                "sample_packing & torch sdpa with bf16 is unsupported may results in 0.0 loss. "
+                "This may work on H100s."
+            )
+
+        return root
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_sample_packing_w_rl(cls, root):
+        if root.get("sample_packing") and root.get("rl"):
+            raise ValueError("`sample_packing: true` does not work with RLHF training")
+        return root
+
+    @model_validator(mode="before")
+    @classmethod
+    def hint_sample_packing_padding(cls, root):
+        if root.get("sample_packing") and not root.get("pad_to_sequence_len"):
+            LOG.warning(
+                "`pad_to_sequence_len: true` is recommended when using sample_packing"
+            )
+        return root
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_gas_bsz(cls, data):
```

Comment:
Is there a reason why some function signatures are `(cls, data)` and some are `(cls, root)`?